dbus: rauc: replace tacd-based update polling with native update polling

hnez opened the pull request:
This uses the native RAUC update polling support introduced in https://github.com/rauc/rauc/pull/1672 to replace the tacd-internal update polling.

Among the various benefits of this approach are the following:

- ETag-based check for changed bundles on the server. This means unchanged bundles do not have to be inspect on every poll.
- Automatic installation and boot into new bundles.

Other PRs related to this one:

- The rauc PR adding the polling feature: https://github.com/rauc/rauc/pull/1672
- The meta-lxatac PR with accompanying configuration: https://github.com/linux-automation/meta-lxatac/pull/242

TODO before un-drafting:

- [x] Write proper commit messages

hnez committed:
dbus: rauc: re-introspect services

RAUC is in the process of adding native update polling support.
Re-introspect the dbus service to add the new APIs.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
dbus: rauc: implement forced polling via update channel files

Users managing a fleet of devices with custom-built update bundles and
update channel may want to automatically enable update polling and
automatic installation of updates without having to do so explicitly via
the web ui. (At least we at Pengutronix do).

Enable this usecase by adding optional `force_polling` and
`force_auto_install` config options to the update channel definition
files.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
dbus: rauc: allow configuring the *_criteria in channel files

The native RAUC polling feature allows configuring when a new update
bundle is even considered as an update candidate (`candidate_criteria`),
when it is considered for automatic installation (`install_criteria`)
and under which conditions to auto-boot into another slot after
installation (`reboot_criteria`).

The defaults we have chosen in previous commits generally make sense,
but allow users with custom update channels to customize them if they
deem it necessary.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
dbus: rauc: reload rauc daemon when required

This triggers RAUC systemd service reload or restart (currently RAUC does
not support reloads, so it will be a restart) and waits for the result.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
dbus: rauc: add support for enabling the auto install feature

Automatic installation and boot of updates can be useful when managing a
fleet of devices. This is however a feature that requires strict user
consent, hence why it is off by default.

Add backend-support for enabling this feature. Frontent support in the
web interface will be added later.

We always enable auto-reboot together with auto-install,
since the migration scripts only run once at the end of the installation.
A system that is updated, but not rebooted, would thus accumulate changes
that are not migrated to the other slot.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
dbus: rauc: add Channels type to use instead of Vec<Channel>

This allows us to add methods like "get the primary channel",
that operates on all update channels at once, to this `Channels`
type later on.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
dbus: rauc: forward poller status to broker

RAUC native polling provides us with information about the recent poll
attempts. This includes information about the bundle version and wether
it is an update over what is currently running on the device.

In other words: it gives us everything we need to show update
notifications again. Forward this information to the same places we used
with the tacd-based update polling.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
dbus: rauc: system_conf: write runtime RAUC config with polling section

This configures RAUC to poll for updates on our behalf.
We do not use the information yet or enable automatic installation but
those are next steps.
We also need to trigger RAUC to re-read the file for this to be useful.
All of these features are added in follow-up commits.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
dbus: rauc: only install bundles from the primary channel

This restricts the sources that the `/v1/tac/update/install` will accept
update requests from to only the primary channel.
The web interface has not exposed the feature to install arbitrary URLs
for some time now and users that want to do so are better served by using
the command line interface instead.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
dbus: rauc: use a UpdateRequest object instead of a simple URL for install

Right now the UpdateRequest object also only contains the URL,
just like before, but in the future we will at least add a
`manifest_hash` field that ensures that the user gets exactly
the update bundle they agreed to install.

The change is backwards-compatible. Incoming requests to the
`/v1/tac/update/install` endpoint are first parsed as JSON object
(the new UpdateRequest object) and if that fails as simple string
(the old URL string) and then transformed transparently.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
dbus: rauc: update_channels: add a concept of a single primary channel

RAUC native update polling only supports a single update channel,
while our native update polling did support multiple
(all channels which RAUC would have accepted updates from,
based on the enabled signing certificates, were polled for updates and the
user was asked if they wanted to install updates from them).

Prepare for the change by adding a concept of a single primary update
channel. The primary channel is the first enabled one. Based on the
channel definition file name.

E.g. on production TACs these channel files are available:

    root@lxatac-00011:~# ls /usr/share/tacd/update_channels/
    01_stable.yaml 05_testing.yaml

They are sorted by name when they are read from disk, so if both
`stable.cert.pem` and `testing.cert.pem` are found
in `/etc/rauc/certificates-enabled/`, then the stable channel will be
the primary channel, but bundles from the testing channel may still be
installed via the command line interface (e.g. to facilitate a channel
switch).

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
dbus: rauc: add manifest_hash und effective_url to UpstreamBundle

The RAUC native polling interface provides more information than
just the basic `compatible` and `version` fields.
Among these extra informations are the following:

  - `manifest_hash`

    By using the `manifest_hash` in the `InstallBundle` call we can
    (cryptographically) ensure that the exact bundle (content) that the
    user agreed to install is actually being installed and that no switch
    has happened in between.

  - `effective_url`

    This is the bundle URL after all HTTP redirects have been followed.
    This is e.g. relevant when a "clever" update server is used that
    redirects poll requests to specific bundles to e.g. implement staged
    rollouts or prevents updates from incompatible bundle versions.

    By using this one can ensure that the bundle URL used matches
    the `manifest_hash` provided and that the redirects did not change
    (e.g. because the next step in a staged update was reached)
    between the last poll and the user accepting an update.

The update dialog on the LCD is updated to use this mechanism now,
while the web interface will be updated later.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
dbus: systemd: also monitor rauc.service

This exposes the status of the RAUC daemon to users inside and outside of
the tacd. As of now neither of them exist, but in the next few commits
support for re-starting the RAUC deamon to reload its configuration will
be added.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
dbus: rauc: allow restricting installation to a specific manifest_hash

This makes sure that an user gets exact the bundle they intended to
install, e.g. the bundle was not replaced by a newer one on the server
or otherwise tampered with.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
dbus: rauc: remove tacd-based update polling

RAUC is in the process of adding native polling support, which we want to
integrate into the tacd.

To do the switch in a reviewable way first remove the tacd-based polling
and then add the native support in separate commits.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
web: setup: inform the user about additional headers that are sent now

We include the `boot-id` and `uptime` options in the RAUC `send-headers`
config in the hopes of detecting boot-loops during update roll-out.
This was not anticipated when writing the setup page, so we add it now.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
web: add toggle switches to enable automatic installation of updates

Only make the toggle visible when polling is enabled, since auto-update
without polling does nothing.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
web: use manifest_hash and effective_url when triggering an install

This ensures that the exact bundle (content) that the user agreed to
install is actually installed.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
web: indicate to the user when a channel is enabled but not primary

The channel list in the web interface now contains a "Upgrade" column
with one of the following:

  - "Not enabled" for channels which are not enabled, which means bundles
    from it can not be installed for it.
  - "Not primary" (this one is new) for channels which are enabled,
    but are not the primary one and are thus not polled by the native RAUC
    polling feature.
  - "Polling disabled" if the polling feature is not enabled.
  - A spinner if we do not know the status yet.
  - "Up to date" if the TAC is in sync with this update channel.
  - "Upgrade" (a button) if an update is available.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
dbus: rauc: trigger a single poll for updates after reloading the daemon

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

jluebbe commented:
I've created a RAUC feature request to support service config reloading: https://github.com/rauc/rauc/issues/1709

jluebbe commented on src/dbus/rauc/system_conf.rs:
```diff
@@ -0,0 +1,120 @@
+    let dynamic_conf = {
+        match poll_section(primary_channel, enable_polling) {
+            Ok(Some(ps)) => {
+                // We use the config in /etc as a template ...
```
You mean the one in `/usr/lib/rauc/`?

hnez replied:
Jep! Changed locally. Thanks!

jluebbe commented on src/dbus/rauc/system_conf.rs:
```diff
@@ -86,8 +87,13 @@ pub fn update_system_conf(
+        // background without them even noticing.
         let polling = enable_polling || force_polling;
-        let auto_install = enable_auto_install || force_auto_install;
+        let auto_install = (enable_auto_install || force_auto_install) && !setup_mode;
```
Couldn't this use the inhibit file mechanism to prevent polling while in setup mode?

hnez replied:
I think there were reasons - like wanting to have up to date information about the upstream bundles even if we are in setup mode.
But I don't think there were strong reasons. I have updated the PR to use the inhibit file mechanism.

jluebbe commented on src/dbus/rauc/system_conf.rs:
```diff
@@ -0,0 +1,161 @@
+    let mut section = String::new();
+
+    writeln!(&mut section)?;
+    writeln!(&mut section, "[poll]")?;
```
We've decided to rename the section from `poll` to `polling`.
```suggestion
    writeln!(&mut section, "[polling]")?;
```
Also update the comment above and in `update_system_conf()`.

hnez committed:
dbus: rauc: also pause auto updating while the DUT power is on

Ideally we would not want to auto-reboot when the TAC is in use,
but deciding when that is, is not easy.
One piece of information we do have is if the DUT is currently powered.
In that case we very likely do not want to reboot right now.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez commented:
I have decided to base this PR on #105 so I do not have to worry about conflicting changes too much.
Since that PR is a draft - because it is currently untested - I will also mark this PR as draft until I had a chance to change that.

hnez committed:
dbus: rauc: break out flatten_slot_status()

The return value from InstallerProxy::get_slot_status() needs some
post-processing.
Break that out into a separate function, because we will need to call
`get_slot_status` from another place soon.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
dbus: rauc: use Err(anyhow!(...)) instead of bail!(...)

I like this better now, because it makes the control flow more obvious
(bail! returns right there, while Err(anyhow!(...))) follows the normal
control flow.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
dbus: rauc: break out booted_slot_other_slot()

This function looks up the two rootfs slots and aranges them based on
which one is booted and which one is not.
This is currently only used in one place, but will be used in another
soon.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
dbus: rauc: mark the booted slot good again after RAUC restart

At least for now the RAUC service will only poll for updates if the booted
slot was marked good during the current runtime of the service.
We restart the service to dynamically update the configuration.
This means we have to mark the booted slot as good again if we want polling.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez commented:
I think we could review and merge this and carry the RAUC changes as patches in `meta-lxatac` for a bit, until we are sure that we like the API and how it works in RAUC.
What do you think @jluebbe?

hnez commented:
We need to get this reviewed. Who could do that? @jluebbe for the RAUC integration and @KarlK90 for the rest?

jluebbe commented on src/dbus/rauc.rs:
```diff
@@ -214,9 +216,43 @@ async fn channel_list_update_task(
+                if ev.active_state == "active" {
+                    break;
+                }
+            }
```
Shouldn't we handle the case where the state doesn't end up as "active" here?

hnez replied:
For example with a timeout?

hnez replied:
I have moved from the `tacd::broker::Topic`  Rube Goldberg machine based service restart to just using the dbus endpoint directly to trigger the restart and get immediate feedback.
This should remove the need for timeouts on our side. Instead the timeouts from the RAUC service file (or the defaults) should take effect.

hnez replied:
It turns out some amount of Rube Goldberg machinery is actually required to reload/restart a systemd service _and be notified about the result_ of this action.
I have just pushed an updated version in the form of https://github.com/linux-automation/tacd/pull/90/changes/6a647400d2559128043a5a81cba52cbc2618723c.

jluebbe commented on src/dbus/rauc/update_channels.rs:
```diff
@@ -176,4 +206,39 @@ impl Channels {
+    }
+
+    #[cfg(not(feature = "demo_mode"))]
+    pub(super) fn update_from_poll_status(&mut self, poll_status: zvariant::Dict) -> Result<bool> {
```
This would benefit from some documentation. How do we find the channel to update from the RAUC status information?

hnez replied:
I will add some text.

hnez replied:
Added some text

jluebbe commented on src/dbus/rauc/system_conf.rs:
```diff
@@ -45,6 +45,10 @@ fn polling_section(
         writeln!(&mut section, "interval-sec={}", interval.as_secs())?;
     }
 
+    let inhibit_files = primary_channel.inhibit_files.as_deref().unwrap_or_default();
```
Why are the inhibit files a per channel configuration? I'd expect that to be system-wide.

hnez replied:
I don't really have a system wide config where I could place this and the update channel configuration files are the closest thing to such a config.

I would assume a strong correlation between users that want to customize the inhibit files and users that have their own update channel.

Users of e.g. the stable update channel are most likely fine with out opinionated default configurations. Users that want to customize will likely need their own update channel configuration.

jluebbe commented on src/dbus/rauc/system_conf.rs:
```diff
@@ -45,7 +45,10 @@ fn polling_section(
+    let inhibit_files = primary_channel
+        .inhibit_files
+        .as_deref()
+        .unwrap_or("/var/run/tacd/inhibit/setup-mode");
```
Perhaps add a tacd-internal inhibit file to the list configured by the user?

hnez replied:
I'd prefer to keep it simple and give the user an option to add or remove inhibit files. Adding one to the config specified by the user would prevent that.
If a user has their own update channel and wants devices to auto-update even if they are in the setup mode, then I think they should be able to.

hnez replied:
The list of inhibit files is not freely configurable anymore. Instead the DUT Power inhibit can now be toggled via the web interface.

hnez commented on src/dbus/rauc/system_conf.rs:
```diff
@@ -0,0 +1,162 @@
+    let inhibit_files = primary_channel
+        .inhibit_files
+        .as_deref()
+        .unwrap_or("/var/run/tacd/inhibit/dut-pwr;/var/run/tacd/inhibit/setup-mode");
```
```suggestion
        .unwrap_or("/run/tacd/inhibit/dut-pwr;/run/tacd/inhibit/setup-mode");
```

jluebbe replied:
Does this cover any existing paths in `/var/run/tacd`?

hnez replied:
I am not sure whether I understand the question. We generate/remove these inhibit files in the tacd and check their existence in RAUC.

hnez commented on src/inhibit.rs:
```diff
@@ -0,0 +1,77 @@
+const VAR_RUN_TACD_INHIBIT: &str = "demo_files/var/run/tacd/inhibit";
+
+#[cfg(not(feature = "demo_mode"))]
+const VAR_RUN_TACD_INHIBIT: &str = "/var/run/tacd/inhibit";
```
```suggestion
const VAR_RUN_TACD_INHIBIT: &str = "demo_files/run/tacd/inhibit";

#[cfg(not(feature = "demo_mode"))]
const VAR_RUN_TACD_INHIBIT: &str = "/run/tacd/inhibit";
```

hnez commented on src/dbus/rauc/update_channels.rs:
```diff
@@ -50,30 +51,52 @@ pub struct Channel {
     pub bundle: Option<UpstreamBundle>,
+    pub force_polling: Option<bool>,
+    pub force_auto_install: Option<bool>,
+    pub inhibit_files: Option<String>,
```
We have decided that we do not want the inhibit file configurability in the channel files.

Instead we want persisted toggles in the web interface for the relevant inhibitors.
"Allow updates while the DUT is powered on" "Allow updates while there are logged in users" (via logind's DBUs interface).

hnez replied:
I have implemented the toggle now. It is hidden behind a "Advanced Settings" expandable section, because it should not be too common to configure it:

<img width="896" height="334" alt="image" src="https://github.com/user-attachments/assets/f0e724ca-0071-4e39-9ca8-10415ddd3ac0" />

<img width="1017" height="389" alt="image" src="https://github.com/user-attachments/assets/8f7da219-555a-4f84-89c3-ad9e068ee67b" />

I think we should implement additional inhibitors in the future. E.g. time based ones or based on the info from `logind`, but first we should get this merged.

hnez committed:
dbus: systemd: make the Service struct remember the unit name

We will need that in the future to implement cleaner service restarts.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
dbus: rauc: split the Rauc struct into config and status topics

The struct itself and also the functions that take parts of it as
parameters (in particular `channel_list_update_task`) are getting
unwieldy.

Adding another parameter to `channel_list_update_task` would cross cargo
clippy's threshold on "too many arguments".

Work around that by splitting the Rauc struct and just passing the config
part of it to `channel_list_update_task` as a whole.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
inhibit: create and remove inhibit file based on setup_mode status

When in setup mode we do not want the system to auto-install updates and
suddenly reboot without user input, as that would lead to a bad first
experience with the TAC.

Instead use an inhibit file to delay the first RAUC update poll to when
the setup mode is exited.

This commit does not yet make use of the inhibit file, it only generates
and removes it.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
inhibit: create/remove an inhibt file based on DUT power state

If a DUT is currently powered by the TAC it is unlikely that we should
reboot for an update.

This commit does not yet make use of the generated inhibit file.
It only creates and removes it based on the DUT power state.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
dbus: rauc: pause auto updating while the TAC is in setup mode

It would be quite surprising (in the negative sense) if the TAC would
reboot without a warning while you are setting it up after unboxing.
Instead wait for the setup to complete before looking for updates.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
web: TacComponents: UpdateChannels: update for single primary channel

With the change to native RAUC update polling only the primary update
channel is checked for updates, not all update channels as before.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
web: MqttComponents: add MqttHider which conditionally hides its children

This allows us to only show some configuration options when a condition is
met.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
web: TacComponents: UpdateConfig: make the DUT Power inhibt configurable

This is an advanced config options that most users should not actually
need, so hide it behind an expandable section (and only show it if polling
is active).

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

hnez committed:
motd: write to /run/tacd/motd instead of /var/run/tacd

The former is the more "modern" path. In practice one will be symlinked
to the other, making this a small change.

Signed-off-by: Leonard Göhrs <l.goehrs@pengutronix.de>

KarlK90 commented on src/dbus/rauc/update_channels.rs:
```diff
@@ -156,85 +193,99 @@ impl Channel {
-        let cert_path = Path::new(ENABLE_DIR).join(cert_file);
-
-        self.enabled = cert_path.exists();
+    pub fn into_vec(self) -> Vec<Channel> {
```
You could also implement `Deref` and `DerefMut` and skip the explicit "unwrapping" of the inner type.

hnez replied:
In this case I actually want to consume `self` and return an owned `Vec` that can be consumed with `into_iter` to get an iterator of owned `Channel`s, so no references anywhere.
This means `Deref` would not be the right choice here.

An alternative to the `channels.into_vec().into_iter()` calls I have here would be to implement [`IntoIterator `](https://doc.rust-lang.org/std/iter/trait.IntoIterator.html) directly:

```rust
impl IntoIterator for Channels {
    type Item = Channel;
    type IntoIter = std::vec::IntoIter;

    fn into_iter(self) -> Self::IntoIter {
        self.0.into_iter()
    }
}
```

But I would prefer to just keep it as it is now.

KarlK90 commented on src/inhibit.rs:
```diff
@@ -0,0 +1,106 @@
+
+        wtb.spawn_task("inhibit-dut-pwr-service", async move {
+            loop {
+                match dut_pwr_state_events.recv().await? {
+                    OutputState::On => self.dut_pwr.inhibit()?,
+                    _ => self.dut_pwr.release()?,
+                }
```
If tacd panics and restarts is there a guarantee that the inhibit files will be cleaned up?

hnez replied:
The files are created / deleted based on the values of the `dut_pwr` and `setup_mode` topics (topics are a Pub/Sub message passing system that the tacd is built around).
Both of them will see a publish during tacd startup¹, resulting in the files being created or deleted (regardless of whether they exist already).
So once the tacd startup completes the existence of the inhibit files will reflect the internal state of the tacd.

---

¹ The starts of the setup_mode topic is read from the state file in `/srv/tacd/state.json` and the `dut_pwr` status defaults to Off.

KarlK90 reviewed (changes requested):
LGTM, just a nit and one question regarding the life cycle of the inhibit files.

hnez commented:
Hi, @KarlK90 and @jluebbe could you have a look at the most recent changes and my review comment feedback?

hnez commented:
I have just rebased the PR on top of the recently merged #107 and a small unmerged fixup commit #108.

I would prefer if we merged this sooner than later, because it should be included in the new tacos release, which I will be preparing next week.
